grafanalib/core.py

```diff
@@ -460,46 +478,60 @@ def _balance_panels(panels):
-    # layer.
-    panels = attr.ib(default=attr.Factory(list), convert=_balance_panels)
-    collapse = attr.ib(
-        default=False, validator=instance_of(bool),
+class Panel(object):
+    gridPos = attr.ib(validator=instance_of(GridPos))
+
+    id = attr.ib(default=None)
+    type = attr.ib(default=None)
+    title = attr.ib(default='')
+    alert = attr.ib(default=None)
+    scopedVars = attr.ib(
+        default=attr.Factory(dict), validator=instance_of(dict)
     )
-    editable = attr.ib(
-        default=True, validator=instance_of(bool),
+    repeat = attr.ib(default=None)
+    repeatIteration = attr.ib(default=None)
+    repeatPanelId = attr.ib(default=None)
+    repeatDirection = attr.ib(default=None)
+    repeatedByRow = attr.ib(default=False, validator=instance_of(bool))
+    maxPerRow = attr.ib(default=None)
+    collapsed = attr.ib(default=False, validator=instance_of(bool))
+    panels = attr.ib(default=attr.Factory(list))
+    soloMode = attr.ib(default=False, validator=instance_of(bool))
+    targets = attr.ib(default=attr.Factory(list), validator=instance_of(list))
+    dataSource = attr.ib(default=None)
+    thresholds = attr.ib(
+        default=attr.Factory(list), validator=instance_of(list)
     )
-    height = attr.ib(
-        default=attr.Factory(lambda: DEFAULT_ROW_HEIGHT),
-        validator=instance_of(Pixels),
+    pluginVersion = attr.ib(default=None)
+
+    snapshotData = attr.ib(
+        default=attr.Factory(list), validator=instance_of(list)
     )
-    showTitle = attr.ib(default=None)
-    title = attr.ib(default=None)
-    repeat = attr.ib(default=None)
+    timeFrom = attr.ib(default=None)
+    timeShift = attr.ib(default=None)
+    hideTimeOverride = attr.ib(default=None)
+    options = attr.ib(default=attr.Factory(list), validator=instance_of(list))
 
-    def _iter_panels(self):
-        return iter(self.panels)
+    maxDataPoints = attr.ib(default=None)
+    interval = attr.ib(default=None)
+    description = attr.ib(default=None)
+    links = attr.ib(default=attr.Factory(list), validator=instance_of(list))
+    transparent = attr.ib(default=False, validator=instance_of(bool))
 
-    def _map_panels(self, f):
-        return attr.assoc(self, panels=list(map(f, self.panels)))
+
+@attr.s
+class Row(Panel):
+    title = attr.ib(default='New row')
 
     def to_json_data(self):
-        showTitle = False
-        title = "New row"
-        if self.title is not None:
-            showTitle = True
-            title = self.title
-        if self.showTitle is not None:
-            showTitle = self.showTitle
         return {
-            'collapse': self.collapse,
-            'editable': self.editable,
-            'height': self.height,
-            'panels': self.panels,
-            'showTitle': showTitle,
-            'title': title,
+            'collapsed': self.collapsed,
+            'gridPos': self.gridPos,
+            'id': self.id,
+            'panels': self.panels if self.collapsed else [],
```

Comment:
I am not sure why this conditional is here, but if Collapsed is not specified or it is set to False, then the JSON dumps has empty panels. I am not sure why you would want this behaviour.

Reply:
I observed this behavior from playing with the Grafana UI and checking the generated JSON. It isn't intuitive to me either, but it seems like this is how Grafana handles collapsed rows.